### PR TITLE
[Live Bidding] App Bidding integration

### PR DESCRIPTION
**Pull request by vguerci**
JIRA: EE-1264

### CODE REVIEW

File: CriteoPublisherSdk/Sources/CR_BidManager.m [hunk 1]
```diff
@@ -288,6 +289,30 @@ - (void)updateTimeToNextCallIfProvided:(CR_CdbResponse *)cdbResponse {
...
+  @try {
+    [self enrichUnsafelyAdObject:object withBid:bid];
+  } @catch (NSException *exception) {
+    CLogException(exception);
+  }
+}
+
+- (void)enrichUnsafelyAdObject:(id)object withBid:(CRBid *)bid {
+  if (!config) {
+    CLog(@"Config hasn't been fetched. So no bids will be fetched.");
```
Comment (TurpIF) [inline CriteoPublisherSdk/Sources/CR_BidManager.m:302]: Bid is already fetched. I guess that such conditions are not valid anymore with the new logic. Same thing for the kill switch.

Moreover, I thought that bidding should not have to wait for the config to be fetched.

File: CriteoPublisherSdk/Tests/IntegrationTests/CR_IntegrationsTestBase.m [hunk 2]
```diff
@@ -39,9 +47,19 @@ - (void)waitForIdleState {
...
 
-- (void)tearDown {
-  [self waitForIdleState];
-  [super tearDown];
+#pragma mark - App Bidding helper
+
+- (void)enrichAdObject:(id)object
+             forAdUnit:(CRAdUnit *)adUnit
+           thenWaitFor:(void (^)(void))completionBlock {
+  [self cr_waitForExpectationBlock:^(XCTestExpectation *expectation) {
```
Comment (TurpIF) [inline CriteoPublisherSdk/Tests/IntegrationTests/CR_IntegrationsTestBase.m:55]: You're doing all your assertions inside a block which make the tests a little bid harder to read (when the block is too big).
What do you think about mocking a synchronous call ? Either via XCTestExpectation or through `[self waitForIdleState]`.

File: CriteoPublisherSdk/Tests/IntegrationTests/CR_ProfileIdFunctionalTests.m [hunk 3]
```diff
@@ -312,4 +327,17 @@ - (NSDictionary *)csmRequest {
   }];
 }
 
+- (void)enrichAd:(id)adObject
+       forAdUnit:(CRAdUnit *)adUnit
+     thenWaitFor:(void (^)(void))completionBlock {
+  [self cr_waitForExpectationBlock:^(XCTestExpectation *expectation) {
```
Comment (TurpIF) [inline CriteoPublisherSdk/Tests/IntegrationTests/CR_ProfileIdFunctionalTests.m:333]: Same comment than in CR_IntegrationTestBase. What do you think about mocking a synchronous call to make test easier to read ?

I see that test are built with `waitForIdleState`. Then tests should already be sync, no ?

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m [hunk 4]
```diff
@@ -406,7 +407,7 @@ - (void)testAddCriteoBidToRequestCallHeaderBidding {
 - (void)testAddCriteoBidToRequestWhenKillSwitchIsEngagedShouldNotEnrichRequest {
```
Comment (TurpIF) [inline CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:407]: I'm not sure that this scenario should be kept. If killSwitch is on, you won't even get a bid. Maybe should you either remove the test, either improve it to test getBid+enrichAdObject.

**Review by TurpIF — CHANGES_REQUESTED** [review]

File: CriteoPublisherSdk/Sources/CR_BidManager.m (hunk 1 above)
Comment (vguerci) [reply CriteoPublisherSdk/Sources/CR_BidManager.m:302]: > Bid is already fetched. I guess that such conditions are not valid anymore with the new logic. Same thing for the kill switch.

For some reason I was thinking the killswitch wasn't meant only for bid requests. Will change this.

> Moreover, I thought that bidding should not have to wait for the config to be fetched.

Not sure how this relates to this particular code as there is no wait there?
Are you talking about the bid fetching general case?

File: CriteoPublisherSdk/Tests/IntegrationTests/CR_IntegrationsTestBase.m (hunk 2 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/IntegrationTests/CR_IntegrationsTestBase.m:55]: This is mostly a way to adapt all tests w/o being too intrusive. While it is maybe not the best one, that ease reviewing, and works with all scenarios, including I think ones that wasn't relying on idleState. I'll see if I can change this w/o too much rewrite / time investment.

File: CriteoPublisherSdk/Tests/IntegrationTests/CR_ProfileIdFunctionalTests.m (hunk 3 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/IntegrationTests/CR_ProfileIdFunctionalTests.m:333]: Same as previous comment, not all tests are based on idle state, this approach works for all scenarios.

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m (hunk 4 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m:407]: Will remove with the logic tested.

File: CriteoPublisherSdk/Sources/CR_BidManager.m (hunk 1 above)
Comment (TurpIF) [reply CriteoPublisherSdk/Sources/CR_BidManager.m:302]: > > Bid is already fetched. I guess that such conditions are not valid anymore with the new logic. Same thing for the kill switch.
> 
> For some reason I was thinking the killswitch wasn't meant only for bid requests. Will change this.

You're right, the kill switch is for deactivating the whole SDK. But IMHO just cutting the network+bidding logic (so `bid=nil`) is enough to deactivate the SDK.
So ok for keeping or removing the killSwitch check.

> 
> > Moreover, I thought that bidding should not have to wait for the config to be fetched.
> 
> Not sure how this relates to this particular code as there is no wait there?
> Are you talking about the bid fetching general case?

I was talking about the `if (!config)`. This condition should not belong to this method as you already have a bid. But I'm even questioning if it belongs to the `loadBid` method: bidding+config are async, so bidding should not care if config was fetched or not.

Comment (vguerci) [reply CriteoPublisherSdk/Sources/CR_BidManager.m:302]: Thanks for the details, will remove then as you said it is covered anyway on bid requesting side which is more central and logical.

File: CriteoPublisherSdk/Tests/IntegrationTests/CR_IntegrationsTestBase.m (hunk 2 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/IntegrationTests/CR_IntegrationsTestBase.m]: So your suggestion works in the end, took a bit more time that I was hoping, but anyway, that's better now, thank you :)

File: CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m (hunk 4 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/UnitTests/CR_BidManagerTests.m]: Done

File: CriteoPublisherSdk/Tests/IntegrationTests/CR_ProfileIdFunctionalTests.m (hunk 3 above)
Comment (vguerci) [reply CriteoPublisherSdk/Tests/IntegrationTests/CR_ProfileIdFunctionalTests.m]: Done

File: CriteoPublisherSdk/Sources/CR_BidManager.m (hunk 1 above)
Comment (vguerci) [reply CriteoPublisherSdk/Sources/CR_BidManager.m]: Done

**Review by TurpIF — APPROVED** [review]

**Pull request merged**